frontend/catalyst/jax_tracer.py

```diff
@@ -541,47 +675,72 @@ def trace_quantum_function(
+            # until there are no more transforms to be applied.
+            # But first we should clean this up this method a bit more.
+            tapes, post_processing = apply_transform(qnode, quantum_tape, return_values_flat)
 
-    abstract_results = tree_unflatten(
-        out_classical_tree, [ShapeDtypeStruct(a.shape, a.dtype, a.named_shape) for a in out_avals]
-    )
-    return closed_jaxpr, abstract_results
+        # (2) - Quantum tracing
+        results_tracers, results_abstract = [], []
+        is_program_transformed = qnode and qnode.transform_program
+        for tape in tapes:
+            # If the program is batched, that means that it was transformed.
+            # If it was transformed, that means that the program might have
+            # changed the output. See `split_non_commuting`
+            if is_program_transformed:
+                # TODO: In the future support arbitrary output from the user function.
+                output = tape.measurements
+                _, trees = jax.tree_util.tree_flatten(output, is_leaf=is_leaf)
+            else:
+                output = return_values_flat
+                trees = return_values_tree
+
+            with EvaluationContext.frame_tracing_context(ctx, trace):
+                qdevice_p.bind(spec="kwargs", val=str(device.backend_kwargs))
+                qdevice_p.bind(spec="backend", val=device.backend_name)
+                qreg_in = qalloc_p.bind(len(device.wires))
+                qrp_out = trace_quantum_tape(tape, device, qreg_in, ctx, trace)
+                meas, meas_trees = trace_quantum_measurements(device, qrp_out, output, trees)
+                out_quantum_tracers = [qrp_out.actualize()]
+                qdealloc_p.bind(qreg_in)
+
+                tracers = [trace.full_raise(m) for m in meas]
+                results_tracers += tracers
+
+                jaxpr, out_type, _ = ctx.frames[trace].to_jaxpr2(tracers + out_quantum_tracers)
+                jaxpr._outvars = jaxpr._outvars[:-1]  # pylint: disable=protected-access
+                out_type = out_type[:-1]
+
+                out_avals, _ = unzip2(out_type)
+                abstract_results = tree_unflatten(
+                    meas_trees,
+                    [ShapeDtypeStruct(a.shape, a.dtype, a.named_shape) for a in out_avals],
+                )
+                # This mimics the return type from qnodes.
+                # I would prefer if qnodes didn't have special rules about whether they return a
+                # tuple, list, or value.
+                # TODO: Allow the user to return whatever types they specify.
+                if is_program_transformed and len(abstract_results) == 1:
+                    results_abstract.append(abstract_results[0])
+                elif is_program_transformed:
+                    results_abstract.append(tuple(abstract_results))
+                else:
+                    results_abstract.append(abstract_results)
```

Comment:
Can we support this now that we do PyTree return values?

Reply:
I think this is more a question about PL and transforms. When I was working on this, I found that the behaviour of the transform changed on whether it was a tuple or a list.

Reply:
Oh I see. Untransformed programs return whatever type the user provided, but since transforms could change the return type arbitrarily (adding results, removing results, changing their type or shape, etc.), we cannot restore the original type tree provided by the user. In this case there is just a standard way to return transform results, which is either a scalar or a tuple.

Sounds good to me 👍